test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,229 @@
+      for (final entry in topicsByStream.entries) {
+        final MapEntry(key: topic, value: sendersByTopic) = entry;
+        for (final entry in sendersByTopic.entries) {
+          final MapEntry(key: senderId, value: tracker) = entry;
+          ((messagesByUserInTopic[streamId] ??= {})[topic] ??= {})
+            [senderId] = Set.from(tracker.ids);
+        }
+      }
+    }
+    return messagesByUserInTopic;
+}
+
+/// [messages] should be sorted by [id] ascendingly.
+void checkMatchesMessages(RecentSenders model, List<Message> messages) {
+  final Map<int, Map<int, Set<int>>> messagesByUserInStream = {};
+  final Map<int, Map<String, Map<int, Set<int>>>> messagesByUserInTopic = {};
+  for (final message in messages) {
+    if (message is! StreamMessage) {
+      throw UnsupportedError('Message of type ${message.runtimeType} is not expected.');
+    }
+
+    final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+
+    ((messagesByUserInStream[streamId] ??= {})
+      [senderId] ??= {}).add(messageId);
+    (((messagesByUserInTopic[streamId] ??= {})[topic] ??= {})
+      [senderId] ??= {}).add(messageId);
+  }
+  final actualMessagesByUserInStream = _streamTrackerToSet(model.streamSenders);
+  final actualMessagesByUserInTopic = _topicTrackerToSet(model.topicSenders);
+  check(actualMessagesByUserInStream).deepEquals(messagesByUserInStream);
+  check(actualMessagesByUserInTopic).deepEquals(messagesByUserInTopic);
+}
+
+void main() {
+  group('RecentSenders', () {
+    test('starts with empty stream and topic senders', () {
+      final model = RecentSenders();
+      checkMatchesMessages(model, []);
+    });
+
+    test('stream message gets included, DM message gets ignored', () {
+      final model = RecentSenders();
+      final streamMessage = eg.streamMessage();
+      final dmMessage = eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]);
+
+      model.handleMessage(streamMessage);
+      model.handleMessage(dmMessage);
+      checkMatchesMessages(model, [streamMessage]);
+    });
+
+    group('add a batch of messages', () {
+      late RecentSenders model;
+      final stream1 = eg.stream(streamId: 1);
+      final user10 = eg.user(userId: 10);
+
+      void setupModel(List<Message> messages) {
+        model = RecentSenders();
+        for (final message in messages) {
+          model.handleMessage(message);
+        }
+      }
+
+      test('batch goes before the existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 100),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 200),
+        ];
```

Comment:
One thing going on in several of these test cases is that the thing they're about (as described by their names) is something that's all about a single tracker at a time; but then they also have some messages that go in a different tracker.

That's two pretty much separate areas of the code being tested, and so the tests are cleanest and easiest to think about if those ideas are covered in separate tests. So these test cases would all be focused on a single tracker; but then there should also be a couple of test cases that show that `handleMessages` correctly handles taking messages that belong in several different trackers, and distributing them across those.

Also that it accepts and ignores DMs — that they don't stop the stream/channel messages in the same batch from getting processed normally.